File: LV 2/삼각 달팽이.py
def solution(n):
    lists = [[0] * i for i in range(1, n+1)]
    idx = 1
    i = 0
    while True:
        x = 2 * i
        y = i
        i += 1
        
        # 아래 내려가는 부분
        while x < n and lists[x][y] == 0:
            lists[x][y] = idx
            x += 1
            idx += 1
        x -= 1
        
        # 오른쪽 진행
        y += 1
        while y < x+1 and lists[x][y] == 0:
            lists[x][y] = idx
            y += 1
            idx += 1
        y -= 1
        
        # 대각선 왼쪽 위 진행
        x -= 1; y -= 1
        while lists[x][y] == 0:
            lists[x][y] = idx
            x -= 1
            y -= 1
            idx += 1
        
        # 전체 체크
        end = True
        for l in lists:
            for e in l:
                if e == 0:
                    end = False
                    break
            if not end:
                break
        
        if end:
            break

    answer = []
    
    # 정답 저장
    for l in lists:
        for e in l:
            answer.append(e)
    return answer

File: LV 2/삼각 달팽이.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int[] solution(int n) {
        List<Integer> answer = new ArrayList<>();
        int[][] matrix = new int[1001][1001];
        int movementCount = n;
        int x = -1, y = 0;
        int number = 1;
        while (true) {
            for (int i = 0; i < movementCount; i++) {
                matrix[++x][y] = number++;
            }
            movementCount--;
            for (int i = 0; i < movementCount; i++) {
                matrix[x][++y] = number++;
            }
            movementCount--;
            for (int i = 0; i < movementCount; i++) {
                matrix[--x][--y] = number++;
            }
            movementCount--;
            if (movementCount < 0) {
                break;
            }
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                answer.add(matrix[i][j]);
            }
        }
        
        return answer.stream().mapToInt(i -> i).toArray();
    }
}
